d2core/d2checkbox/checkbox.go

```diff
@@ -0,0 +1,155 @@
+package d2checkbox
+
+import (
+	"github.com/gravestench/akara"
+
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2geom/rectangle"
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2interface"
+	"github.com/OpenDiablo2/OpenDiablo2/d2common/d2resource"
+	"github.com/OpenDiablo2/OpenDiablo2/d2core/d2label"
+)
+
+// set up defaults
+const (
+	CheckboxDefaultTextOffset = 18
+	CheckboxDefaultWidth      = 16
+	CheckboxDefaultHeight     = 15
+)
+
+type callbackFunc = func(this akara.Component) (preventPropagation bool)
+
+// Checkbox defines a standard wide UI button
+type Checkbox struct {
+	Layout   CheckboxLayout
+	Sprite   d2interface.Sprite
+	Label    *d2label.Label
+	callback callbackFunc
+	pressed  bool
+	enabled  bool
+}
+
+// CheckboxLayout defines the type of buttons
+type CheckboxLayout struct {
+	X                float64
+	Y                float64
+	SpritePath       string
+	PalettePath      string
+	FontPath         string
+	ClickableRect    *rectangle.Rectangle
+	XSegments        int
+	YSegments        int
+	BaseFrame        int
+	DisabledFrame    int
+	DisabledColor    uint32
+	TextOffset       float64
+	FixedWidth       int
+	FixedHeight      int
+	LabelColor       uint32
+	Toggleable       bool
+	AllowFrameChange bool
+	HasImage         bool
+	Tooltip          int
+	TooltipXOffset   int
+	TooltipYOffset   int
+}
+
+// New creates an instance of Button
+func New() *Checkbox {
+	checkbox := &Checkbox{
+		Layout: GetDefaultLayout(),
+	}
+
+	return checkbox
+}
+
+// GetDefaultLayout returns the default layout of a checkbox.
+func GetDefaultLayout() CheckboxLayout {
+	return CheckboxLayout{
+		X:                0,
+		Y:                0,
+		SpritePath:       d2resource.Checkbox,
+		PalettePath:      d2resource.PaletteFechar,
+		FontPath:         d2resource.FontExocet10,
+		XSegments:        1,
+		YSegments:        1,
+		BaseFrame:        0,
+		DisabledFrame:    -1,
+		DisabledColor:    lightGreyAlpha75,
+		TextOffset:       CheckboxDefaultTextOffset,
+		FixedWidth:       CheckboxDefaultWidth,
+		FixedHeight:      CheckboxDefaultHeight,
+		LabelColor:       goldAlpha100,
+		Toggleable:       true,
+		AllowFrameChange: true,
+		HasImage:         true,
+		Tooltip:          0,
+		TooltipXOffset:   0,
+		TooltipYOffset:   0,
+	}
+}
+
+// OnActivated defines the callback handler for the activate event
+func (v *Checkbox) OnActivated(callback callbackFunc) {
+	v.callback = callback
+}
+
+// Activate calls the on activated callback handler, if any
+func (v *Checkbox) Activate(thisComponent akara.Component) bool {
+	if v.GetEnabled() {
+		v.Toggle()
+	}
+
+	if v.callback != nil {
+		return v.callback(thisComponent)
+	}
+
+	return false
+}
```

Comment:
Why does this method need an argument?
would it make more sense to pass `v` to `v.callback` as the reference to `thisComponent` ?
```golang
	if v.callback != nil {
		return v.callback(v)
	}
```

Reply:
In general, I think it would be useful to have access to the object itself in your callback. I think this is how a lot of callback  and event-based stuff in Javascript works.

I think it would be a lot cleaner looking to just pass `v` in to the callback, but that does not give developers as much control, since `v` is the 'inner' `d2checkbox.Checkbox` type and not a `d2component.Checkbox`. The latter (the way it is currently written) gives you the EID of the checkbox so there's more that you are able to do within the callback. I think that is preferable, but it's definitely kind of dirty looking.

Pending your thoughts, if we want to continue passing the outer component in, we could move `Activate` and some of these other functions to the outer `d2component.Checkbox` (and do that for other components as well, like buttons), and that would give us the ability to easily pass it into the callback without weird hacky workarounds.

Reply:
passing the EID around makes great sense, good point!

I don't really have any issues with the code as it is, i just wanted to discuss.